visitors/http_rest_diff/http_rest_spec_diff_visitor.go

```diff
@@ -0,0 +1,550 @@
+	EnterChangedHTTPAuth(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.HTTPAuth) Cont
+	LeaveAddedOrRemovedHTTPAuth(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.HTTPAuth, cont Cont) Cont
+	LeaveChangedHTTPAuth(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.HTTPAuth, cont Cont) Cont
+
+	EnterAddedOrRemovedData(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Data) Cont
+	LeaveAddedOrRemovedData(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Data, cont Cont) Cont
+
+	EnterAddedOrRemovedList(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.List) Cont
+	LeaveAddedOrRemovedList(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.List, cont Cont) Cont
+
+	EnterAddedOrRemovedOneOf(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.OneOf) Cont
+	LeaveAddedOrRemovedOneOf(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.OneOf, cont Cont) Cont
+
+	EnterAddedOrRemovedOptional(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Optional) Cont
+	LeaveAddedOrRemovedOptional(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Optional, cont Cont) Cont
+
+	EnterAddedOrRemovedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive) Cont
+	EnterChangedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive) Cont
+	LeaveAddedOrRemovedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive, cont Cont) Cont
+	LeaveChangedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive, cont Cont) Cont
+
+	EnterAddedOrRemovedStruct(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Struct) Cont
+	LeaveAddedOrRemovedStruct(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Struct, cont Cont) Cont
+}
+
+// An HttpRestSpecDiffVisitor with convenience functions for entering and
+// leaving nodes with diffs.
+type DefaultHttpRestSpecDiffVisitor interface {
+	HttpRestSpecDiffVisitor
+
+	EnterDiff(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
+	LeaveDiff(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}, cont Cont) Cont
+
+	// Delegates to EnterDiff by default.
+	EnterAddedOrRemovedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
+
+	// Delegates to EnterDiff by default.
+	EnterChangedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
```

Comment:
So, I thought about this some more, and there is another option not yet discussed.  But since you're doing type assertions anyway it should be about equal cost (they are cheap in Go.)

Instead of creating a large interface, and then a base class to provide sensible defaults, you could create small interfaces. Then move that decision about what method to call elsewhere, so that instead of EnterAddedOrRemovedDiff having a "default" implementation, the visit method sees whether the visitor has the appropriate interface--- and if not, drop down to a simpler one. 

https://play.golang.org/p/FkRz48Zfjxe

I'm not 100% sure this is a fit, but it seems to solve this tactical problem of "how do I delegate within the interface" by moving that delegation to something that already has the original object.

You can even do this as an adaption layer, you could have one big interface made of lots of the small interfaces, and all the "what do I delegate to" logic lives within that.

Reply:
We already manually dispatch on the type of the node being visited: https://github.com/akitasoftware/akita-libs/blob/jed/diff-visitor/visitors/http_rest/http_rest_spec_pair_visitor.go#L436. I'm not convinced that additionally dispatching on the visitor type for each of those cases is going to be better.

I'm also inclined to leave things as they are, in the name of "scrappiness"; especially if the only thing gained from a refactoring is the badge of being idiomatic.